wire.md

```diff
@@ -0,0 +1,1008 @@
+particular user.
+
+#### 6.2.5 `post/topic`
+
+Set a topic for a channel.
+
+field          | type               | desc
+---------------|--------------------|-------------------------------------------------------
+`channel_len`  | `varint`           | length of the channel's name, in bytes
+`channel`      | `u8[channel_len] ` | channel name (UTF-8)
+`topic_len`    | `varint`           | length of the topic field, in bytes
+`topic`        | `u8[topic_len] `   | topic content (UTF-8)
+
+`post_type` MUST be set to `3`.
+
+A `topic` field MUST be a valid UTF-8 string, between 0 and 512 codepoints. A
+topic of length zero MUST be considered as the current topic being cleared to
+the empty string, "".
+
+#### 6.2.6 `post/join`
+
+Publicly announce membership in a channel.
+
+field          | type               | desc
+---------------|--------------------|-------------------------------------------------------
+`channel_len`  | `varint`           | length of the channel's name, in bytes
+`channel`      | `u8[channel_len] ` | channel name (UTF-8)
+
+`post_type` MUST be set to `4`.
+
+#### 6.2.7 `post/leave`
+
+Publicly announce termination of membership in a channel.
+
+field          | type               | desc
+---------------|--------------------|-------------------------------------------------------
+`channel_len`  | `varint`           | length of the channel's name, in bytes
+`channel`      | `u8[channel_len] ` | channel name (UTF-8)
+
+`post_type` MUST be set to `5`.
+
+### 6.3 Message Formats
+
+#### 6.3.1 Message Header
+
+All messages MUST begin with the following header fields:
+
+field         | type     | desc
+--------------|----------|-------------------------------------------------------------
+`msg_len`     | `varint` | number of bytes in rest of message, not including the `msg_len` field
+`msg_type`    | `varint` | a type identifier for the message, which controls which fields follow this header
+`circuit_id`  | `u8[4]`  | id of a circuit for an established path, or `[0,0,0,0]` for no circuit
+`req_id`      | `u8[4]`  | unique id of this request (random)
+
+Message-specific fields follow after the `req_id`.
+
+Each request and response type has a unique `msg_type` (see below), which
+controls which fields will immediately follow this header.
+
+Clients encountering a `msg_type` they do not know how to parse MUST ignore and
+discard it.
+
+The `circuit_id` field is not currently in use, and MUST be set to all zeros.
+
+The request ID, `req_id`, is a 32-bit number, generated randomly by the
+requester. It is used to uniquely identify the request during its lifetime
+across the peers who may handle it. For networks with up to 43 million active
+requests within a single cabal, the probability of collision remains below 1%.
+(i.e. `(1 - 1 / (2**32)) ** 43_000_000 > 0.99`)
+
+When generating a new `req_id`, a client MUST discard and generate a new one if
+the one generated collides with a known active `req_id`.
```

Comment:
Don't you trust the collision probability you just computed? If that probability is low enough, you should not need to perform this quite expensive check (expensive because I MUST track all known active ids [ignore this point if I already MUST track them for other reasons]). If you need to do this check, then the collision probability is too high.

Reply:
I redid the math, using the birthday problem to calculate the probabilities, and realized that 2**32 is not enough. I've changed `req_id` to be 64-bit instead.

My new math shows that, given 10,000 active hosts sending 1 request/second, there'd only be a 1% chance of collision per year, which I feel a lot better about. Even if there is a collision, then the consequence is that some peers may drop that incoming request, believing it to be one they have already seen.

Thanks for commenting on this!